source/components/inputs/typeaheadList/typeaheadList.ts

```diff
@@ -0,0 +1,180 @@
+import { TypeaheadDataItemComponent } from './typeaheadDataItem';
+import { ListHeaderTemplate, ListItemTemplate } from './templates/index';
+
+export interface ITypeaheadListChanges {
+	value: SimpleChange;
+	disableSearching: SimpleChange;
+	[key: string]: SimpleChange;
+}
+
+@Component({
+	selector: 'rlTypeaheadList',
+	template: require('./typeaheadList.html'),
+	inputs: validationInputs,
+	outputs: baseOutputs,
+	providers: [ComponentValidator],
+	directives: [TypeaheadComponent, SelectComponent, ButtonAsyncComponent, TypeaheadDataItemComponent],
+})
+export class TypeaheadListComponent<T> extends ValidatedInputComponent<T[]> implements OnInit, OnChanges {
+	@Input() transform: __transform.ITransform<T, string>;
+	@Input() getItems: { (search?: string): Promise<T[]> | Observable<T[]> };
+	@Input() prefix: string;
+	@Input() clientSearch: boolean;
+	@Input() disableSearching: boolean;
+	@Input() create: { (value: string): T };
+	@Input() onAdd: { (item: T): Promise<T> | Observable<T> };
+	@Input() onRemove: { (item: T): Promise<void> | Observable<void> };
+	@Output() select: EventEmitter<T> = new EventEmitter<T>();
+
+	@ContentChild(ListHeaderTemplate) listHeader: ListHeaderTemplate;
+	@ContentChild(ListItemTemplate) listItem: ListItemTemplate;
+
+	cachedItemsArray: T[];
+	cachedItems: BehaviorSubject<T[]>;
+
+	transformService: __transform.ITransformService;
+	searchUtility: __search.ISearchUtility;
+
+	constructor(transformService: __transform.TransformService
+			, @Optional() rlForm: FormComponent
+			, componentValidator: ComponentValidator
+			, object: __object.ObjectUtility
+			, array: __array.ArrayUtility
+			, guid: __guid.GuidService
+			, searchUtility: __search.SearchUtility) {
+		super(rlForm, componentValidator, object, array, guid);
+		this.transformService = transformService;
+		this.searchUtility = searchUtility;
+		this.inputType = 'typeaheadList';
+		this.value = [];
+	}
+
+	loadItems(search?: string): Observable<T[]> {
+		if (this.clientSearch || this.disableSearching) {
+			if (this.cachedItems) {
+				return this.cachedItems;
+			} else {
+				return Observable.from(this.getItems());
+			}
+		} else {
+			return Observable.from(this.getItems(search));
+		}
+	}
+
+	searchItems = (search?: string): Observable<T[]> => {
+		return this.loadItems(search).map((items: T[]): T[] => {
+			return this.filter(items, search);
+		});
+	}
+
+	add = (item: T): Observable<T> => {
+		const action = this.asObservable(this.onAdd(item));
+		action.subscribe(newItem => {
+			newItem = newItem || item;
+			// immutability?
+			const newValue = clone(this.value);
+			newValue.push(newItem);
+			this.setValue(newValue);
+			if (this.cachedItems) {
+				this.array.remove(this.cachedItemsArray, item);
+				this.cachedItems.next(this.cachedItemsArray);
+			}
+			return newItem;
+		});
+		return action;
+	}
+
+	remove = (item: T): Observable<void> => {
+		const action = this.asObservable(this.onRemove(item));
+		action.subscribe(() => {
```

Comment:
Sure, we could do that. I usually pair review these bigger PRs with someone anyway